src/Symfony/Component/Semaphore/Store/RedisStore.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Semaphore\Store;
+
+use Symfony\Component\Cache\Traits\RedisClusterProxy;
+use Symfony\Component\Cache\Traits\RedisProxy;
+use Symfony\Component\Semaphore\Exception\InvalidArgumentException;
+use Symfony\Component\Semaphore\Exception\SemaphoreAcquiringException;
+use Symfony\Component\Semaphore\Exception\SemaphoreExpiredException;
+use Symfony\Component\Semaphore\Key;
+use Symfony\Component\Semaphore\PersistingStoreInterface;
+
+/**
+ * RedisStore is a PersistingStoreInterface implementation using Redis as store engine.
+ *
+ * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RedisStore implements PersistingStoreInterface
+{
+    private $redis;
+
+    /**
+     * @param \Redis|\RedisArray|\RedisCluster|\Predis\ClientInterface $redisClient
+     */
+    public function __construct($redisClient)
+    {
+        if (!$redisClient instanceof \Redis && !$redisClient instanceof \RedisArray && !$redisClient instanceof \RedisCluster && !$redisClient instanceof \Predis\ClientInterface && !$redisClient instanceof RedisProxy) {
+            throw new InvalidArgumentException(sprintf('%s() expects parameter 1 to be Redis, RedisArray, RedisCluster or Predis\ClientInterface, %s given', __METHOD__, \is_object($redisClient) ? \get_class($redisClient) : \gettype($redisClient)));
+        }
+
+        $this->redis = $redisClient;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(Key $key, float $ttlInSecond)
+    {
+        $script = '
+            local key = KEYS[1]
+            local counterKey = key .. ":counter"
+            local ownerKey = key .. ":owner"
+            local timeKey = key .. ":time"
+            local identifier = ARGV[1]
+            local now = tonumber(ARGV[2])
+            local ttlInSecond = tonumber(ARGV[3])
+            local limit = tonumber(ARGV[4])
+
+            -- Remove expired values
+            redis.call("ZREMRANGEBYSCORE", timeKey, "-inf", now)
+            redis.call("ZINTERSTORE", ownerKey, 2, ownerKey, timeKey, "WEIGHTS", 1, 0)
+
+            -- Try to get a semaphore
+            local counter = redis.call("INCR", counterKey)
```

Comment:
What happens when the key become to big? Could it be reset somehow when ownerKey is empty?.

Reply:
you are talking of *VERY BIG NUMBER* here :) 
9223372036854775807 or (2 power 63 -1)
I'm not sure this is an issue.

But indeed we could reset the conter when the buckets are empty.
I think I will also expire the 3 keys also

I will clean other parts of the code for now (+ add tests).  But I keep this comment open.

Reply:
I fixed this issue with a TTL + a reset of the key when all other bucket are empty